torch/_decomp/decompositions_for_rng.py

```diff
@@ -0,0 +1,270 @@
+            throw_if_philox_offset_did_not_advance(fn)
+        )
+
+    return inner
+
+
+def throw_on_non_cuda(device):
+    raise RuntimeError(
+        f"You are trying to functionalize a {device.type} RNG operator but {device.type} does not "
+        f"use Philox/counter-based PRNG. Therefore, functionalizing a {device.type} RNG operator is "
+        "not supported. We are discussing the possibility of a Philox-based RNG implementation for CPU."
+    )
+
+
+def rand_offset_calculator(shape):
+    # For impl, look at the function calc_execution_policy in the file
+    # aten/src/ATen/native/cuda/DistributionTemplates.h. The impl was copied at
+    # commit hash 72aa0667bd16707d50eb8fa337092a1f5d11dfb6
+    numel = 1
+    for dim_size in shape:
+        numel *= dim_size
+
+    block_size = 256
+    unroll = 4
+    curand4_engine_calls = 4
+    device_property = torch.cuda.get_device_properties(torch.cuda.current_device())
+    blocks_per_sm = int(device_property.max_threads_per_multi_processor / block_size)
+    grid_size = int((numel + block_size - 1) / block_size)
+    grid_size = min(grid_size, device_property.multi_processor_count * blocks_per_sm)
+    offset = (
+        int((numel - 1) / (block_size * grid_size * unroll) + 1) * curand4_engine_calls
+    )
+    return offset
+
+
+# TODO - We have to register many more distributions here, and also higher level
+# ops like dropout which have fused implementation and can hide the rand inside.
+@register_rng_decomposition(aten.rand)
+def rand(shape, dtype=None, layout=torch.strided, device=None, pin_memory=False):
+    device = device or "cpu"
+
+    if device.type != "cuda":
+        throw_on_non_cuda(device)
+    seed, offset = PhiloxStateTracker.get_state_as_tuple()
+    dtype = dtype or torch.float32
+    stride = make_contiguous_strides_for(shape)
+    r = torch.ops.rngprims.philox_rand(shape, seed, offset, stride, device, dtype)
+    PhiloxStateTracker.advance_offset(rand_offset_calculator(shape))
+    return r
+
+
+@register_rng_decomposition(aten.rand_like)
+def rand_like(
+    x: torch.Tensor,
+    dtype=None,
+    layout=None,
+    device=None,
+    pin_memory=False,
+    memory_format=torch.preserve_format,
+):
+    device = device or x.device
+    if device.type != "cuda":
+        throw_on_non_cuda(device)
+    dtype = dtype or x.dtype
+    seed, offset = PhiloxStateTracker.get_state_as_tuple()
+    r = torch.ops.rngprims.philox_rand(x.shape, seed, offset, x.stride(), device, dtype)
+    PhiloxStateTracker.advance_offset(rand_offset_calculator(x.shape))
+    return r
+
+
+class PhiloxState:
+    """
+    Represents a PhiloxRngState - (seed, offset) where offset = base_offset +
+    relative_offset. seed and base_offset basically point to the rng state just
+    before tracing starts. relative offset tracks the totaly consumed offset at
+    trace time.
+    """
+
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
+        self.seed = torch.Tensor(())
```

Comment:
note that my suggestion was the lower case constructor: torch.tensor(())
This is the one we recommend over the Tensor() one these days.